go/vt/vttablet/tabletmanager/replmanager.go

```diff
@@ -0,0 +1,185 @@
+// fixed. On state change, SetTabletType must be called before changing
+// the tabletserver state. This will ensure that replication is fixed
+// upfront, allowing tabletserver to start off healthy.
+type replManager struct {
+	ctx          context.Context
+	tm           *TabletManager
+	markerFile   string
+	ticks        *timer.Timer
+	firstFailure bool
+
+	// replStopped is tri-state.
+	// A nil value signifies that the value is not set.
+	mu          sync.Mutex
+	replStopped *bool
+}
+
+func newReplManager(ctx context.Context, tm *TabletManager, interval time.Duration) *replManager {
+	return &replManager{
+		ctx:        ctx,
+		tm:         tm,
+		markerFile: markerFile(tm.Cnf),
+		ticks:      timer.NewTimer(interval),
+	}
+}
+
+func (rm *replManager) SetTabletType(tabletType topodatapb.TabletType) {
+	if *mysqlctl.DisableActiveReparents {
+		return
+	}
+	if !topo.IsReplicaType(tabletType) {
+		rm.ticks.Stop()
+		return
+	}
+	if rm.replicationStopped() {
+		// Stop just to be safe.
+		rm.ticks.Stop()
+		log.Info("Replication Manager: stopped")
+		return
+	}
+	if rm.ticks.Running() {
+		return
+	}
+	log.Info("Replication Manager: starting")
+	// Run an immediate check to fix replication if it was broken.
+	// A higher caller may already have te action lock. So, we use
+	// a code path that avoids it.
+	rm.checkActionLocked()
+	rm.ticks.Start(rm.check)
+}
+
+func (rm *replManager) check() {
+	// We need to obtain the action lock if we're going to fix
+	// replication
+	if err := rm.tm.lock(rm.ctx); err != nil {
+		return
+	}
+	defer rm.tm.unlock()
+	rm.checkActionLocked()
+}
+
+func (rm *replManager) checkActionLocked() {
+	status, err := rm.tm.MysqlDaemon.ReplicationStatus()
+	if err != nil {
+		if err != mysql.ErrNotReplica {
+			return
+		}
+	} else {
+		// If only one of the threads is stopped, it's probably
+		// intentional. So, we don't repair replication.
+		if status.SQLThreadRunning || status.IOThreadRunning {
+			return
+		}
+	}
+
+	if rm.firstFailure {
+		log.Infof("Replication is stopped, reconnecting to master.")
+	}
+	ctx, cancel := context.WithTimeout(rm.ctx, 5*time.Second)
+	defer cancel()
+	if err := rm.tm.repairReplication(ctx); err != nil {
+		if rm.firstFailure {
+			rm.firstFailure = false
+			log.Infof("Failed to reconnect to master: %v, will keep retrying.", err)
+		}
+		return
+	}
+	log.Info("Successfully reconnected to master.")
+	rm.firstFailure = true
+}
+
+// setReplicationStopped performs a best effort attempt of
+// remembering a decision to stop replication.
+func (rm *replManager) setReplicationStopped(stopped bool) {
+	if *mysqlctl.DisableActiveReparents {
+		return
+	}
+
+	rm.mu.Lock()
+	defer rm.mu.Unlock()
+
+	rm.replStopped = &stopped
+
+	if stopped {
+		rm.ticks.Stop()
+	} else {
+		rm.ticks.Start(rm.check)
+	}
+
+	if rm.markerFile == "" {
+		return
+	}
+	if stopped {
+		if file, err := os.Create(rm.markerFile); err == nil {
+			file.Close()
+		}
+	} else {
+		os.Remove(rm.markerFile)
+	}
+}
```

Comment:
When markerFile be empty? If it is only for test, can we change the test to not have empty markerFile?

Reply:
`markerFile` can be empty if the mysql is unmanaged. There are end2end tests for testing the overall behavior. I've written some unit tests to cover the rest of the corner cases that are not covered by the e2e tests.